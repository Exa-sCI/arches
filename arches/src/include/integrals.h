#pragma once
#include <algorithm>
#include <memory>

// Abstract storage object for 1 and 2 electron integrals
template <class T> class JChunk {

  protected:
    std::unique_ptr<idx_t[]> idx_ptr;
    std::unique_ptr<T[]> J_ptr;

  public:
    idx_t size; // size of chunk
    idx_t *ind; // compound two- or four- indices
    T *J;       // integrals

    template <class Y> JChunk(idx_t N, idx_t *J_ind, Y *J_val) {
        size = N;
        std::unique_ptr<idx_t[]> p_ind(new idx_t[size]);
        std::unique_ptr<Y[]> p_val(new Y[size]);

        idx_ptr = std::move(p_ind);
        J_ptr = std::move(p_val);

        std::copy(J_ind, J_ind + size, ind);
        std::copy(J_val, J_val + size, J);
    }

    ~JChunk() = default;
};
