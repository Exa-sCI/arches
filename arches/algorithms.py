# ruff : noqa : E741

from typing import Iterable

import numpy as np
from mpi4py import MPI
from scipy.linalg import lapack

from arches.fundamental_types import Determinant
from arches.integrals import JChunk
from arches.matrix import AMatrix, diagonalize


def bmgs_h(X, p, Q_0=None, R_0=None, T_0=None):
    """
    Block version of Modified Gram-Schmidt orthogonalization
    as described in Barlow, 2019 (https://doi.org/10.1137/18M1197400)

    Relates MGS via Househoulder QR to pose MGS
    via MM operations instead of VV or MV products

    TODO: Handle case when n // p != n / p, i.e., number of columns not divisible by block size
    Should be able to cleanly pad by zeros

    TODO: Possible to write in-place version?

    Args:
        X (array)   : m by n matrix to orthogonalize
        p (int)     : size of block
        Q_0 (array) : m by (s*p) matrix, initial s*p orthogonal vectors
        R_0 (array) : (s*p) by (s*p) matrix, initial R matrix s.t. X = Q @ R
        T_0 (array) : (s*p) by (s*p) matrix, initial T matrix s.t. T = (triu(Q.T @ Q))^-1
    """

    m, n = X.shape
    s = n // p

    # preallocate Q, R, and T
    Q = np.zeros_like(X)
    R = np.zeros((n, n))
    T = np.zeros((n, n))

    if (Q_0 is None) or (R_0 is None) or (T_0 is None):
        Q_0, R_0 = np.linalg.qr(X[:, :p], mode="reduced")
        T_0 = np.eye(p)
        Q[:, :p] = Q_0
        R[:p, :p] = R_0
        T[:p, :p] = T_0
        s_idx = 1
    else:
        # infer start step from size of input guesses
        s_idx = Q_0.shape[1] // p
        ss = slice(int(s_idx * p))
        Q[:, ss] = Q_0
        R[ss, ss] = R_0
        T[ss, ss] = T_0

    for k in range(s_idx, s):
        a = slice(k * p)
        b = slice(k * p, (k + 1) * p)

        H_k = Q[:, a].T @ X[:, b]
        H_k = T[a, a].T @ H_k

        Y_k = X[:, b] - Q[:, a] @ H_k
        Q_kk, R_kk = np.linalg.qr(Y_k, mode="reduced")  # any fast, stable QR can replace this
        F_k = Q[:, a].T @ Q_kk
        G_k = -T[a, a] @ F_k

        Q[:, b] = Q_kk

        R[a, b] = H_k
        R[b, b] = R_kk

        T[a, b] = G_k
        T[b, b] = np.eye(p)

    return Q, R, T


def davidson(H, *args, V_0=None, N_states=1, l=32, pc="D", max_iter=100, tol=1e-6, **kwargs):  # noqa: E741
    """
    Implementation of the Davidson diagonalization algorithm in serial,
    with blocked eigenvector prediction and BMGS.
    """
    if V_0 is None:
        V_k = np.zeros((H.shape[0], l))
        V_k[:l, :l] = np.eye(l)
    else:
        V_k = V_0

    R_k = np.eye(l)
    T_k = np.eye(l)
    N = H.shape[0]
    C_d = np.diag(H)
    if pc == "T":
        C_sd = np.zeros(N - 1)
        for i in range(N - 1):
            C_sd[i] = H[i, i + 1]

        ptsvx = lapack.get_lapack_funcs("ptsvx", dtype=H.dtype)

    for k in range(max_iter):
        ### Project Hamiltonian onto subspace spanned by V_k
        W_k = H @ V_k
        S_k = V_k.T @ W_k

        ### diagonalize S_k to find (projected) eigenbasis
        L_k, Q_k = diagonalize(S_k)  # L_k is sorted smallest -> largest

        ### Form Ritz vectors (in place) and calculate residuals
        R_k = V_k @ Q_k[:, :l] @ L_k[:l, :l] - W_k @ Q_k[:, :l]

        r_norms = np.linalg.norm(R_k, axis=0)

        # process early exits
        early_exit = False
        if np.all(r_norms[:N_states] < tol):
            early_exit = True
            print(f"Davidson has converged with tolerance {tol:0.4e}")
            for i in range(N_states):
                print(f"Energy of state {i}:{L_k[i]} Ha")

        elif k == max_iter - 1:
            early_exit = True
            print("Davidson has hit max number of iterations. Exiting early.")
            for i in range(N_states):
                print(f"Energy of state {i}:{L_k[i]} Ha. Residual {r_norms[i]}")

        if early_exit:
            return L_k[:N_states], Q_k[:, :N_states]

        ### Calculate next batch of trial vectors
        # TODO: get rid of allocatons in loop
        V_kk = np.zeros((H.shape[0], l))

        for i in range(l):
            if pc == "D":  # diagonal preconditioner
                C_ki = np.ones(N) * L_k[i] - C_d
                C_ki = 1.0 / C_ki
                V_kk[:, i] = C_ki * R_k[:, i]
            elif pc == "T":  # tridiagonal preconditioner
                C_d_i = np.ones(N) * L_k[i] - C_d
                res = ptsvx(C_d_i, C_sd, R_k[:, i])
                V_kk[:, i] = res[0]

        ### Form new orthonormal subspace
        V_k, R_k, T_k = bmgs_h(np.hstack([V_k, V_kk]), l // 2, V_k, R_k, T_k)


def davidson_par_0(
    H, comm, *args, V_0=None, N_states=1, l=32, pc="D", max_iter=100, tol=1e-6, **kwargs
):  # noqa: E741
    """
    Implementation of the Davidson diagonalization algorithm in parallel.

    Assumes each worker i has partial component of H s.t. H = \sum_i H_i

    Assume that subspace is small enough that diagonalization/BMGS are fast enough locally
    to outspeed communicating
    """

    if V_0 is None:
        V_k = np.zeros((H.shape[0], l))
        V_k[:l, :l] = np.eye(l)
    else:
        V_k = V_0

    R_k = np.eye(l)
    T_k = np.eye(l)
    N = H.shape[0]

    ## TODO: Diagonal and subdiagonal need to be communicated
    C_d = np.diag(H)
    if pc == "T":
        C_sd = np.zeros(N - 1)
        for i in range(N - 1):
            C_sd[i] = H[i, i + 1]

        ptsvx = lapack.get_lapack_funcs("ptsvx", dtype=H.dtype)

    rank = comm.Get_rank()
    assert l == comm.Get_size()

    for k in range(max_iter):
        ### Project Hamiltonian onto subspace spanned by V_k
        W_k = H @ V_k
        comm.Allreduce(MPI.IN_PLACE, [W_k, MPI.DOUBLE])  ## Synchronization

        S_k = V_k.T @ W_k

        ### diagonalize S_k to find (projected) eigenbasis
        # assume S_k \in R^(kl, kl) is small enough to diagonalize locally
        L_k, Q_k = diagonalize(S_k)  # L_k is sorted smallest -> largest

        ### Calculate residual of owned Ritz vector
        R_k = L_k[rank] * V_k @ Q_k[:, rank] - W_k @ Q_k[:, rank]

        r_norm = np.linalg.norm(R_k)
        r_norms = np.zeros(l)
        comm.Allgather([r_norm, MPI.DOUBLE], [r_norms, MPI.DOUBLE])  ## Synchronization

        # process early exits
        early_exit = False
        if np.all(r_norms[:N_states] < tol):
            early_exit = True
            if rank == 0:
                print(f"Davidson has converged with tolerance {tol:0.4e}")
                for i in range(N_states):
                    print(f"Energy of state {i}:{L_k[i]} Ha")
        elif k == max_iter - 1:
            early_exit = True
            if rank == 0:
                print("Davidson has hit max number of iterations. Exiting early.")
                for i in range(N_states):
                    print(f"Energy of state {i}:{L_k[i]} Ha. Residual {r_norms[i]}")

        if early_exit:
            return L_k[:N_states], Q_k[:, :N_states]

        ### Calculate next batch of trial vectors
        # TODO: get rid of allocatons in loop
        V_kk = np.zeros((H.shape[0], l))

        if pc == "D":  # diagonal preconditioner
            C_ki = np.ones(N) * L_k[rank] - C_d
            C_ki = 1.0 / C_ki
            V_kk[:, rank] = C_ki * R_k
        elif pc == "T":  # tridiagonal preconditioner
            C_d_i = np.ones(N) * L_k[rank] - C_d
            res = ptsvx(C_d_i, C_sd, R_k)
            V_kk[:, rank] = res[0]

        comm.Allgather(MPI.IN_PLACE, [V_kk, MPI.DOUBLE])

        ### Form new orthonormal subspace
        V_k, R_k, T_k = bmgs_h(np.hstack([V_k, V_kk]), l // 2, V_k, R_k, T_k)


def get_connected_dets(int_dets: Iterable[Determinant]) -> Iterable[Determinant]:
    raise NotImplementedError


def cipsi(
    E0,
    int_dets,
    psi_coef,
    J_chunks: Iterable[JChunk],
    *args,
    pt2_threshold=1e-8,
    exc_constraints=None,
    **kwargs,
):
    ext_dets = get_connected_dets(int_dets, exc_constraints)

    e_pt2_n = np.array(len(ext_dets))  # TODO: convert these arrays to LinkedArrays
    e_pt2_d = np.array(len(ext_dets))  # TODO: initialize with E0

    for chunk in J_chunks:
        kernels = (
            chunk.pt2_kernels
        )  # kernels[0] is numerator contrib., kernels[1] is denom. contrib.
        match chunk.category:
            case "OE" | "F":  # both num and denominator
                kernels[0](chunk, int_dets, psi_coef, ext_dets, e_pt2_n)
                kernels[1](chunk, ext_dets, e_pt2_d, E0)
            case "A" | "B":  # only denominator
                kernels[1](chunk, ext_dets, e_pt2_d, E0)
            case _:  # only numerator
                kernels[0](chunk, int_dets, psi_coef, ext_dets, e_pt2_n)

    # reduce e_pt2_n,d over processes
    e_pt2 = e_pt2_n / e_pt2_d  # TODO: implement +, -, /, * for LinkedArrays
    pt2_filter = e_pt2 > pt2_threshold  # TODO: pass this off to customizable selection function
    e_pt2_total = np.sum(e_pt2)

    # TODO: sort ext dets so that they are in weight order? Or otherwise, pass in information if N_max_dets will be hit
    return int_dets + ext_dets[pt2_filter], e_pt2_total


def prepare_hamiltonian(psi_dets, H_partial: AMatrix = None) -> AMatrix:
    raise NotImplementedError


def s_CI(
    E0,
    dets,
    psi_coef,
    J_chunks,
    *args,
    N_states=1,
    N_max_dets=1e6,
    pt2_conv=1e-4,
    **kwargs,
):
    est_pt2_correction = np.inf

    while len(dets) < N_max_dets and est_pt2_correction > pt2_conv:
        # expand wavefunction in space of Slater determinants
        dets, est_pt2_correction = cipsi(E0, dets, psi_coef, J_chunks)

        # find wavefunctions and energies of first N states
        H = prepare_hamiltonian(dets)
        E0, psi_coef = davidson(H)

    return E0, psi_coef, dets
