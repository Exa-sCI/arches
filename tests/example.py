import pathlib

import numpy as np

from arches.algorithms import evaluate_H_entries, s_CI
from arches.determinant import det_t, spin_det_t
from arches.integrals import load_integrals_into_chunks
from arches.io import load_integrals
from arches.linked_object import LinkedArray_idx_t, get_LArray
from arches.log_timer import logtimer
from arches.matrix import DMatrix, SymCSRMatrix


class FakeComm:
    # to avoid initializing MPI for just tests
    def __init__(self, rank, size):
        self.rank = rank
        self.size = size

    def Get_rank(self):
        return self.rank

    def Get_size(self):
        return self.size


def get_ground_state(N_orbs, N_elec):
    max_orb = N_elec // 2
    sdet = spin_det_t(N_elec, occ=True, max_orb=max_orb)
    return det_t(N_orbs, sdet, sdet)


def get_E0(ground_state, chunks, V_nn, dtype):
    A_p = LinkedArray_idx_t(2, np.array([0, 1]).astype(np.int64))
    A_c = LinkedArray_idx_t(1, np.array([0]).astype(np.int64))
    A_v = get_LArray(dtype=dtype)(1)
    H = SymCSRMatrix(1, 1, dtype, A_p=A_p.arr, A_c=A_c.arr, A_v=A_v.arr)

    evaluate_H_entries(H, ground_state, chunks, V_nn)
    return H.A_v[0]


def get_constraint(ground_det, N_holes, N_parts):
    occ_orbs = ground_det.alpha.as_orb_list
    all_orbs = set([x for x in range(ground_det.N_orbs)])
    h_constraint = occ_orbs[-N_holes:]
    p_constraint = tuple(all_orbs.difference(occ_orbs).difference(set(h_constraint)))[:N_parts]
    return (h_constraint, p_constraint)


config_map = (
    lambda fp, config, **kwargs: (str(fp).rsplit("/", 1)[1], config["constraint"]),
    lambda name, constraint: f"| Example: {name.split('.')[0]}, (N_h, N_p): {constraint}",
)


@logtimer(config_map, report_on=("call",))
def run_example(fp, config, dtype):
    N_orbs, N_elec = load_integrals(str(fp), return_size_only=True)
    ground_det = get_ground_state(N_orbs, N_elec)
    ground_psi = DMatrix(1, 1, 1.0, dtype=dtype)
    config["constraint"] = get_constraint(ground_det, *config["constraint"])

    N_orbs, N_elec, V_nn, chunks = load_integrals_into_chunks(
        str(fp),
        FakeComm(0, 1),
        dtype=dtype,
        constraint=config["constraint"],
        chunk_size=512,
        screening_threshold=config["screening_threshold"],
    )

    E0 = get_E0(ground_det, chunks, V_nn, dtype)

    (E_var, psi_var, dets) = s_CI(V_nn, E0, ground_det, ground_psi, chunks, **config)


def nh3(dtype):
    print("###### Running NH3 example ######\n\n")
    fp = pathlib.Path("../data/nh3.5det.fcidump")
    config = {
        "N_states": 1,
        "N_max_dets": 2000,
        "pt2_conv": 1e-6,
        "pt2_threshold": 1e-8,
        "constraint": (4, 4),
        "screening_threshold": 1e-8,
    }
    run_example(fp, config, dtype)
    print("###### NH3 example finished ######\n")


def f2(dtype):
    print("###### Running F2 example ######")
    fp = pathlib.Path("../data/f2_631g.18det.fcidump")
    config = {
        "N_states": 1,
        "N_max_dets": 2000,
        "pt2_conv": 1e-6,
        "pt2_threshold": 1e-8,
        "constraint": (4, 4),
        "screening_threshold": 1e-8,
    }
    run_example(fp, config, dtype)
    print("###### F2 example finished ######\n")


def c2(dtype):
    print("###### Running C2 example ######")
    fp = pathlib.Path("../data/c2_eq_hf_dz.fcidump")
    config = {
        "N_states": 1,
        "N_max_dets": 2000,
        "pt2_conv": 1e-6,
        "pt2_threshold": 1e-8,
        "constraint": (4, 4),
        "screening_threshold": 1e-8,
    }
    run_example(fp, config, dtype)
    print("###### C2 example finished ######\n")


def main(dtype=np.float64):
    nh3(dtype)
    print("------------------\n")
    f2(dtype)
    print("------------------\n")
    c2(dtype)


if __name__ == "__main__":
    main()
