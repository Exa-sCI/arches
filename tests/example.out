###### Running NH3 example ######


Starting selected CI routine with 1 determinants. SCF energy -56.1610 Ha

###################

Generating connected space for 1 internal determinants | Execution took 2.586e-04 seconds
Calculating pt2 contribution for 117 external determinants | Execution took 3.617e-02 seconds
Internal det space is now rank 48
CIPSI with 1 internal determinants | Execution took 3.676e-02 seconds
Estimated pt2 contribution -3.3192e-01 Ha
Finding non-zero Hamiltonian entries | Execution took 2.600e-04 seconds
Calculating Hamiltonian entries | Execution took 2.522e-01 seconds
Generating Hamiltonian with rank 48 | Execution took 2.525e-01 seconds
Davidson has converged with tolerance 1.00e-06 in 7 iterations
Energy of state 0 : -5.617241e+01 Ha. Residual 5.910570e-07
Davidson for sparse H with 550 entries | Execution took 4.043e-02 seconds

## sCI Iteration Summary
E0 after 1 rounds (Ha): E0_var : -56.1724 | E_pt2 : -3.3192e-01  | E_var + pt2 : -56.5043

###################

Generating connected space for 48 internal determinants | Execution took 1.426e-03 seconds
Calculating pt2 contribution for 333 external determinants | Execution took 3.910e+00 seconds
Internal det space is now rank 175
CIPSI with 48 internal determinants | Execution took 3.912e+00 seconds
Estimated pt2 contribution 6.9314e-01 Ha
Finding non-zero Hamiltonian entries | Execution took 1.532e-03 seconds
Calculating Hamiltonian entries | Execution took 1.264e+00 seconds
Generating Hamiltonian with rank 175 | Execution took 1.265e+00 seconds
Davidson has converged with tolerance 1.00e-06 in 18 iterations
Energy of state 0 : -5.776000e+01 Ha. Residual 5.870659e-07
Davidson for sparse H with 4575 entries | Execution took 3.893e-02 seconds

## sCI Iteration Summary
E0 after 2 rounds (Ha): E0_var : -57.7600 | E_pt2 : 6.9314e-01  | E_var + pt2 : -57.0669

###################

Generating connected space for 175 internal determinants | Execution took 1.946e-03 seconds
Calculating pt2 contribution for 225 external determinants | Execution took 9.779e+00 seconds
Internal det space is now rank 208
CIPSI with 175 internal determinants | Execution took 9.782e+00 seconds
Estimated pt2 contribution 1.8001e-02 Ha
Finding non-zero Hamiltonian entries | Execution took 2.069e-03 seconds
Calculating Hamiltonian entries | Execution took 1.771e+00 seconds
Generating Hamiltonian with rank 208 | Execution took 1.773e+00 seconds
Davidson has converged with tolerance 1.00e-06 in 24 iterations
Energy of state 0 : -5.776038e+01 Ha. Residual 8.032140e-07
Davidson for sparse H with 6424 entries | Execution took 4.727e-02 seconds

## sCI Iteration Summary
E0 after 3 rounds (Ha): E0_var : -57.7604 | E_pt2 : 1.8001e-02  | E_var + pt2 : -57.7424

###################

Generating connected space for 208 internal determinants | Execution took 1.820e-03 seconds
Calculating pt2 contribution for 192 external determinants | Execution took 9.887e+00 seconds
Internal det space is now rank 208
CIPSI with 208 internal determinants | Execution took 9.889e+00 seconds
Estimated pt2 contribution 8.5221e-12 Ha
Finding non-zero Hamiltonian entries | Execution took 2.083e-03 seconds
Calculating Hamiltonian entries | Execution took 1.776e+00 seconds
Generating Hamiltonian with rank 208 | Execution took 1.778e+00 seconds
Davidson has converged with tolerance 1.00e-06 in 24 iterations
Energy of state 0 : -5.776038e+01 Ha. Residual 8.032140e-07
Davidson for sparse H with 6424 entries | Execution took 4.554e-02 seconds

## sCI Iteration Summary
E0 after 4 rounds (Ha): E0_var : -57.7604 | E_pt2 : 8.5221e-12  | E_var + pt2 : -57.7604
run_example ran with config | Example: nh3, (N_h, N_p): (3, 3) and executed in 2.891e+01 seconds
###### NH3 example finished ######

------------------

###### Running F2 example ######
Starting selected CI routine with 1 determinants. SCF energy -198.6461 Ha

###################

Generating connected space for 1 internal determinants | Execution took 1.698e-04 seconds
Calculating pt2 contribution for 117 external determinants | Execution took 5.419e-02 seconds
Internal det space is now rank 48
CIPSI with 1 internal determinants | Execution took 5.454e-02 seconds
Estimated pt2 contribution -5.4869e-01 Ha
Finding non-zero Hamiltonian entries | Execution took 1.596e-04 seconds
Calculating Hamiltonian entries | Execution took 3.622e-01 seconds
Generating Hamiltonian with rank 48 | Execution took 3.624e-01 seconds
Davidson has converged with tolerance 1.00e-06 in 8 iterations
Energy of state 0 : -1.987388e+02 Ha. Residual 8.977276e-07
Davidson for sparse H with 553 entries | Execution took 9.516e-03 seconds

## sCI Iteration Summary
E0 after 1 rounds (Ha): E0_var : -198.7388 | E_pt2 : -5.4869e-01  | E_var + pt2 : -199.2875

###################

Generating connected space for 48 internal determinants | Execution took 9.628e-04 seconds
Calculating pt2 contribution for 333 external determinants | Execution took 9.606e+00 seconds
Internal det space is now rank 95
CIPSI with 48 internal determinants | Execution took 9.608e+00 seconds
Estimated pt2 contribution 5.1484e-02 Ha
Finding non-zero Hamiltonian entries | Execution took 5.207e-04 seconds
Calculating Hamiltonian entries | Execution took 9.850e-01 seconds
Generating Hamiltonian with rank 95 | Execution took 9.856e-01 seconds
Davidson has converged with tolerance 1.00e-06 in 12 iterations
Energy of state 0 : -1.987408e+02 Ha. Residual 6.985353e-07
Davidson for sparse H with 1488 entries | Execution took 2.005e-02 seconds

## sCI Iteration Summary
E0 after 2 rounds (Ha): E0_var : -198.7408 | E_pt2 : 5.1484e-02  | E_var + pt2 : -198.6894

###################

Generating connected space for 95 internal determinants | Execution took 1.122e-03 seconds
Calculating pt2 contribution for 305 external determinants | Execution took 1.775e+01 seconds
Internal det space is now rank 96
CIPSI with 95 internal determinants | Execution took 1.775e+01 seconds
Estimated pt2 contribution -8.7196e-04 Ha
Finding non-zero Hamiltonian entries | Execution took 5.146e-04 seconds
Calculating Hamiltonian entries | Execution took 1.066e+00 seconds
Generating Hamiltonian with rank 96 | Execution took 1.067e+00 seconds
Davidson has converged with tolerance 1.00e-06 in 14 iterations
Energy of state 0 : -1.987408e+02 Ha. Residual 3.817094e-07
Davidson for sparse H with 1536 entries | Execution took 2.994e-02 seconds

## sCI Iteration Summary
E0 after 3 rounds (Ha): E0_var : -198.7408 | E_pt2 : -8.7196e-04  | E_var + pt2 : -198.7417

###################

Generating connected space for 96 internal determinants | Execution took 1.205e-03 seconds
Calculating pt2 contribution for 304 external determinants | Execution took 1.857e+01 seconds
Internal det space is now rank 96
CIPSI with 96 internal determinants | Execution took 1.857e+01 seconds
Estimated pt2 contribution 1.5710e-09 Ha
Finding non-zero Hamiltonian entries | Execution took 5.157e-04 seconds
Calculating Hamiltonian entries | Execution took 1.074e+00 seconds
Generating Hamiltonian with rank 96 | Execution took 1.074e+00 seconds
Davidson has converged with tolerance 1.00e-06 in 14 iterations
Energy of state 0 : -1.987408e+02 Ha. Residual 3.817094e-07
Davidson for sparse H with 1536 entries | Execution took 3.111e-02 seconds

## sCI Iteration Summary
E0 after 4 rounds (Ha): E0_var : -198.7408 | E_pt2 : 1.5710e-09  | E_var + pt2 : -198.7408
run_example ran with config | Example: f2_631g, (N_h, N_p): (3, 3) and executed in 4.960e+01 seconds
###### F2 example finished ######

------------------

###### Running C2 example ######
Starting selected CI routine with 1 determinants. SCF energy -75.3869 Ha

###################

Generating connected space for 1 internal determinants | Execution took 1.750e-04 seconds
Calculating pt2 contribution for 117 external determinants | Execution took 1.294e-01 seconds
Internal det space is now rank 50
CIPSI with 1 internal determinants | Execution took 1.298e-01 seconds
Estimated pt2 contribution -2.7000e+01 Ha
Finding non-zero Hamiltonian entries | Execution took 1.864e-04 seconds
Calculating Hamiltonian entries | Execution took 1.045e+00 seconds
Generating Hamiltonian with rank 50 | Execution took 1.045e+00 seconds
Davidson has converged with tolerance 1.00e-06 in 17 iterations
Energy of state 0 : -7.551950e+01 Ha. Residual 7.927081e-07
Davidson for sparse H with 600 entries | Execution took 2.343e-02 seconds

## sCI Iteration Summary
E0 after 1 rounds (Ha): E0_var : -75.5195 | E_pt2 : -2.7000e+01  | E_var + pt2 : -102.5194

###################

Generating connected space for 50 internal determinants | Execution took 1.126e-03 seconds
Calculating pt2 contribution for 331 external determinants | Execution took 2.410e+01 seconds
Internal det space is now rank 97
CIPSI with 50 internal determinants | Execution took 2.410e+01 seconds
Estimated pt2 contribution -2.7111e-01 Ha
Finding non-zero Hamiltonian entries | Execution took 5.338e-04 seconds
Calculating Hamiltonian entries | Execution took 2.643e+00 seconds
Generating Hamiltonian with rank 97 | Execution took 2.643e+00 seconds
Davidson has converged with tolerance 1.00e-06 in 20 iterations
Energy of state 0 : -7.553678e+01 Ha. Residual 5.414403e-07
Davidson for sparse H with 1535 entries | Execution took 3.811e-02 seconds

## sCI Iteration Summary
E0 after 2 rounds (Ha): E0_var : -75.5368 | E_pt2 : -2.7111e-01  | E_var + pt2 : -75.8079

###################

Generating connected space for 97 internal determinants | Execution took 1.258e-03 seconds
Calculating pt2 contribution for 303 external determinants | Execution took 4.385e+01 seconds
Internal det space is now rank 98
CIPSI with 97 internal determinants | Execution took 4.385e+01 seconds
Estimated pt2 contribution 1.9125e-02 Ha
Finding non-zero Hamiltonian entries | Execution took 8.590e-04 seconds
Calculating Hamiltonian entries | Execution took 2.835e+00 seconds
Generating Hamiltonian with rank 98 | Execution took 2.836e+00 seconds
Davidson has converged with tolerance 1.00e-06 in 20 iterations
Energy of state 0 : -7.553748e+01 Ha. Residual 9.959066e-07
Davidson for sparse H with 1583 entries | Execution took 4.725e-02 seconds

## sCI Iteration Summary
E0 after 3 rounds (Ha): E0_var : -75.5375 | E_pt2 : 1.9125e-02  | E_var + pt2 : -75.5184

###################

Generating connected space for 98 internal determinants | Execution took 1.472e-03 seconds
Calculating pt2 contribution for 302 external determinants | Execution took 4.465e+01 seconds
Internal det space is now rank 98
CIPSI with 98 internal determinants | Execution took 4.466e+01 seconds
Estimated pt2 contribution 2.2094e-07 Ha
Finding non-zero Hamiltonian entries | Execution took 6.625e-04 seconds
Calculating Hamiltonian entries | Execution took 2.849e+00 seconds
Generating Hamiltonian with rank 98 | Execution took 2.850e+00 seconds
Davidson has converged with tolerance 1.00e-06 in 20 iterations
Energy of state 0 : -7.553748e+01 Ha. Residual 9.959066e-07
Davidson for sparse H with 1583 entries | Execution took 4.510e-02 seconds

## sCI Iteration Summary
E0 after 4 rounds (Ha): E0_var : -75.5375 | E_pt2 : 2.2094e-07  | E_var + pt2 : -75.5375
run_example ran with config | Example: c2_eq_hf_dz, (N_h, N_p): (3, 3) and executed in 1.225e+02 seconds
###### C2 example finished ######

